File: src/erlang/ch01/1.4/l1/src/int_arith.erl
%% Integer arithmetic policies: checked / wrapping / saturating (i32/i64) + checked mul.
-module(int_arith).
-export([
    add_i32_checked/2, add_i32_wrapping/2, add_i32_saturating/2, mul_i32_checked/2,
    add_i64_checked/2, add_i64_wrapping/2, add_i64_saturating/2, mul_i64_checked/2
]).

-define(INT32_MIN, -2147483648).
-define(INT32_MAX,  2147483647).
-define(INT64_MIN, -9223372036854775808).
-define(INT64_MAX,  9223372036854775807).

%% --- helpers ---

to_signed_bits(Value, Bits) ->
    Mask = (1 bsl Bits) - 1,
    V = Value band Mask,
    Sign = 1 bsl (Bits - 1),
    case (V band Sign) =/= 0 of
        true  -> V - (1 bsl Bits);
        false -> V
    end.

wrap_add_bits(A, B, Bits) ->
    to_signed_bits((A + B) band ((1 bsl Bits) - 1), Bits).

%% --- i32 ---

add_i32_checked(A, B) ->
    S = A + B,
    if S < ?INT32_MIN orelse S > ?INT32_MAX -> {error, overflow};
       true                                 -> {ok, S}
    end.

add_i32_wrapping(A, B) ->
    wrap_add_bits(A, B, 32).

add_i32_saturating(A, B) ->
    S = A + B,
    if S > ?INT32_MAX -> ?INT32_MAX;
       S < ?INT32_MIN -> ?INT32_MIN;
       true           -> S
    end.

mul_i32_checked(A, B) ->
    P = A * B,
    if P < ?INT32_MIN orelse P > ?INT32_MAX -> {error, overflow};
       true                                 -> {ok, P}
    end.

%% --- i64 ---

add_i64_checked(A, B) ->
    S = A + B,
    if S < ?INT64_MIN orelse S > ?INT64_MAX -> {error, overflow};
       true                                 -> {ok, S}
    end.

add_i64_wrapping(A, B) ->
    wrap_add_bits(A, B, 64).

add_i64_saturating(A, B) ->
    S = A + B,
    if S > ?INT64_MAX -> ?INT64_MAX;
       S < ?INT64_MIN -> ?INT64_MIN;
       true           -> S
    end.

mul_i64_checked(A, B) ->
    P = A * B,
    if P < ?INT64_MIN orelse P > ?INT64_MAX -> {error, overflow};
       true                                 -> {ok, P}
    end.

File: src/c/ch01/1.4/l1/int_arith.c
#include "int_arith.h"

/* ---------- int32 ---------- */

bool add_i32_checked(int32_t a, int32_t b, int32_t *out) {
    if ((b > 0 && a > INT32_MAX_C - b) ||
        (b < 0 && a < INT32_MIN_C - b)) {
        return false;
    }
    if (out) *out = (int32_t)(a + b);
    return true;
}

int32_t add_i32_wrapping(int32_t a, int32_t b) {
    uint32_t ua = (uint32_t)a, ub = (uint32_t)b;
    return (int32_t)(ua + ub); /* two's complement wrap */
}

int32_t add_i32_saturating(int32_t a, int32_t b) {
    if ((b > 0 && a > INT32_MAX_C - b)) return INT32_MAX_C;
    if ((b < 0 && a < INT32_MIN_C - b)) return INT32_MIN_C;
    return (int32_t)(a + b);
}

bool mul_i32_checked(int32_t a, int32_t b, int32_t *out) {
    /* widen to int64 for safe check */
    int64_t p = (int64_t)a * (int64_t)b;
    if (p > INT32_MAX_C || p < INT32_MIN_C) return false;
    if (out) *out = (int32_t)p;
    return true;
}

/* ---------- int64 ---------- */

bool add_i64_checked(int64_t a, int64_t b, int64_t *out) {
    if ((b > 0 && a > INT64_MAX_C - b) ||
        (b < 0 && a < INT64_MIN_C - b)) {
        return false;
    }
    if (out) *out = a + b;
    return true;
}

int64_t add_i64_wrapping(int64_t a, int64_t b) {
    uint64_t ua = (uint64_t)a, ub = (uint64_t)b;
    return (int64_t)(ua + ub); /* two's complement wrap */
}

int64_t add_i64_saturating(int64_t a, int64_t b) {
    if ((b > 0 && a > INT64_MAX_C - b)) return INT64_MAX_C;
    if ((b < 0 && a < INT64_MIN_C - b)) return INT64_MIN_C;
    return a + b;
}

bool mul_i64_checked(int64_t a, int64_t b, int64_t *out) {
    /* Portable overflow test via division bounds (no __int128 needed). */
    if (a == 0 || b == 0) { if (out) *out = 0; return true; }
    if (a > 0) {
        if (b > 0) { if (a > INT64_MAX_C / b) return false; }
        else       { if (b < INT64_MIN_C / a) return false; }
    } else {
        if (b > 0) { if (a < INT64_MIN_C / b) return false; }
        else       { if (a != 0 && b < INT64_MAX_C / a) return false; }
    }
    if (out) *out = a * b;
    return true;
}
